File: ReverseALinkedList/ReverseLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 * 1-2-3-4-5-6-null
 * 6-5-4-3-2-1-null
 */

//Iterative solution
class ReverseLinkedList {
	public ListNode reverseList(ListNode head) {

		if(head == null || head.next == null) return head;
		ListNode current = head, temp = null;

		while(current.next != null) {
			temp = current.next;
			current.next = temp.next;
			temp.next = head;
			head = temp;
		}

		return head;

	}
}

//Recursive solution
class ReverseLinkedList {
	public ListNode reverseList(ListNode head) {
		if(head == null || head.next == null) return head;
		return helper(head, head);
	}

	ListNode helper(ListNode head, ListNode current) {
		//Base case
		if(current.next == null) return head;
		ListNode temp = null;

		temp = current.next;
		current.next = temp.next;
		temp.next = head;
		head = temp;

		return helper(head, current);
	}
}

File: reverse-linked-list/reverse-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {
        if(head==null || head.next ==null)
        {
            return head;
        }
        ListNode curr = head;
        ListNode prev = null;
        ListNode temp = head.next;
        while(temp!=null)
        {
            curr.next = prev;
            prev = curr;
            curr = temp;
            temp = temp.next;
            
            
        }
        curr.next =prev;
        return curr;
        
    }
}
